kernel/kernel-api/src/main/java/io/delta/kernel/internal/commit/CommitContextImpl.java

```diff
@@ -0,0 +1,155 @@
+import io.delta.kernel.data.Row;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.actions.CommitInfo;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.transaction.TransactionDataSource;
+import io.delta.kernel.internal.transaction.TransactionV2State;
+import io.delta.kernel.internal.util.VectorUtils;
+import io.delta.kernel.utils.CloseableIterator;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class CommitContextImpl implements CommitContext {
+
+  ////////////////////////////
+  // Static factory methods //
+  ////////////////////////////
+
+  /**
+   * Creates a {@link CommitContext} to be used for the first commit attempt of a transaction only
+   * (e.g. not a retry, which requires conflict detection and potential action reconciliation and
+   * rebasing).
+   *
+   * @param finalizedDataActions the finalized data actions to be committed. That is, all updates to
+   *     the data actions (e.g. row tracking updates) have been applied. Note that these are the
+   *     data actions, not the metadata actions (like CommitInfo, Protocol, Metadata,
+   *     SetTransaction, DomainMetadata, etc.).
+   */
+  public static CommitContextImpl forFirstCommitAttempt(
+      Engine engine, TransactionV2State txnState, CloseableIterator<Row> finalizedDataActions) {
+    return new CommitContextImpl(engine, txnState, finalizedDataActions);
+  }
+
+  // TODO: forRebasedCommit
+
+  //////////////////////////////////
+  // Member variables and methods //
+  //////////////////////////////////
+
+  private final TransactionV2State txnState;
+  private final CloseableIterator<Row> finalizedDataActions;
+  private final long commitAttemptTimestampMs;
+  private final CommitInfo commitInfo;
+  private final Metadata metadata;
+  private boolean iteratorConsumed;
+
+  private CommitContextImpl(
+      Engine engine, TransactionV2State txnState, CloseableIterator<Row> finalizedDataActions) {
+    this.txnState = txnState;
+    this.finalizedDataActions = finalizedDataActions;
+    this.commitAttemptTimestampMs = txnState.clock.getTimeMillis();
+    this.metadata = txnState.initialMetadata; // TODO: update with ICT enablement info on conflict
+    this.commitInfo = getCommitInfo();
+    this.iteratorConsumed = false;
+  }
+
+  /////////////////
+  // Public APIs //
+  /////////////////
+
+  @Override
+  public CloseableIterator<Row> getFinalizedActions() {
+    checkState(!iteratorConsumed, "Finalized actions iterator has already been consumed.");
+    iteratorConsumed = true;
+    return getMetadataActions().combine(finalizedDataActions);
+  }
+
+  @Override
+  public CommitMetadata getCommitMetadata() {
+    return new CommitMetadata(
+        getCommitAsVersion(),
+        txnState.logPath,
+        commitInfo,
+        txnState.readTableOpt.map(TransactionDataSource::getProtocol),
+        txnState.readTableOpt.map(TransactionDataSource::getMetadata),
```

Comment:
why not `txnState.initialMetadata`?

Reply:
> Is this because we don't have intent based APIs, so if we're rebasing we may find existing properties that weren't there before and have to update them?

Good question. No that's not the case / why. As far as I know, we only update the metadata between commit attempts because of In-Commit-Timestamp requirements.

We need to set delta table properties in the commit that enables ICT. These properties include the enablement timetamp and version.

So if we want to write commit 10, we will use an enablement version of 10.

If we conflict with 10.json, we will then try again and write 11.json. But now in this commit attempt we need to update the ICT enablement version property to be 11.